matrix_sdk/src/async_client.rs

```diff
@@ -745,11 +765,89 @@ impl AsyncClient {
                 }
             }
         }
+        Ok(updated)
+    }
 
-        let mut client = self.base_client.write().await;
-        client.receive_sync_response(&mut response, updated).await?;
+    #[inline]
+    async fn iter_left_rooms(&self, response: &mut sync_events::Response) -> Result<bool> {
+        let mut updated = false;
+        for (room_id, left_room) in &mut response.rooms.leave {
+            for mut event in &mut left_room.timeline.events {
+                let decrypted_event = {
+                    let mut client = self.base_client.write().await;
 
-        Ok(response)
+                    let (decrypt_ev, timeline_update) = client
+                        .receive_joined_timeline_event(room_id, &mut event)
+                        .await;
+                    if timeline_update {
+                        updated = true;
+                    };
+                    decrypt_ev
+                };
+
+                if let Some(e) = decrypted_event {
+                    *event = e;
+                }
+
+                if let Ok(e) = event.deserialize() {
+                    let client = self.base_client.read().await;
+                    client.emit_timeline_event(&room_id, &e).await;
+                }
+            }
+
+            // re looping is not ideal here
+            for event in &mut left_room.state.events {
+                if let Ok(e) = event.deserialize() {
+                    let client = self.base_client.read().await;
+                    client.emit_state_event(&room_id, &e).await;
+                }
+            }
+
+            if updated {
+                if let Some(store) = self.base_client.read().await.state_store.as_ref() {
+                    let mut client = self.base_client.write().await;
+                    let room = client.get_or_create_room(&room_id).clone();
```

Comment:
This creates a new Room inside the `joined_rooms` hashmap. How do users that get events emitted from an invited/left room differentiate if the room is in any of the 3 forms?

Reply:
Should `BaseClient` have `joined_rooms`, `left_rooms` and `invited_rooms`, then When the events are emitted the `Room` could be wrapped in an enum 
```rust
pub trait EventEmitter {
    on_room_*(&self, room: RoomState<Arc<RwLock<Room>>>
...

pub enum RoomState {
   Invited(...),
   Joined(...),
   Left(...),
}
```
When a room is joined or left it could be moved into the corresponding HashMap?

Reply:
Yeah I think an enum should be fine. And yeah, moving the rooms to their hashmaps should be done.